File: frontend/src/types/questionGeneration.ts
/**
 * STEP 33: Constrained Question Generator types.
 */
export type GenerationStatus = 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED' | 'CANCELLED';
export type ValidationStatus = 'PENDING' | 'PASSED' | 'PASSED_WITH_WARNINGS' | 'FAILED';
export type ReviewStatus = 'DRAFT' | 'REVIEWED' | 'APPROVED' | 'REJECTED';
export type GenQuestionType = 'mcq' | 'short_answer' | 'descriptive' | 'problem_solving' | 'true_false' | 'conceptual' | 'analytical';
export type GenDifficulty = 'easy' | 'medium' | 'hard';
export type GenCognitive = 'Remember' | 'Understand' | 'Apply' | 'Analyze' | 'Evaluate' | 'Create';
export type FeedbackReason = 'too_easy' | 'too_difficult' | 'wrong_topic' | 'wrong_co' | 'too_similar' | 'poor_wording' | 'not_appropriate' | 'other';

export const QUESTION_TYPE_LABELS: Record<GenQuestionType, string> = {
  mcq: 'Multiple choice', short_answer: 'Short answer', descriptive: 'Descriptive', problem_solving: 'Problem solving',
  true_false: 'True / False', conceptual: 'Conceptual', analytical: 'Analytical',
};
export const DIFFICULTY_LEVELS: GenDifficulty[] = ['easy', 'medium', 'hard'];
export const COGNITIVE_LEVELS: GenCognitive[] = ['Remember', 'Understand', 'Apply', 'Analyze', 'Evaluate', 'Create'];
export const FEEDBACK_REASONS: { value: FeedbackReason; label: string }[] = [
  { value: 'too_easy', label: 'Too easy' }, { value: 'too_difficult', label: 'Too difficult' }, { value: 'wrong_topic', label: 'Wrong topic' },
  { value: 'wrong_co', label: 'Wrong course outcome' }, { value: 'too_similar', label: 'Too similar to existing questions' },
  { value: 'poor_wording', label: 'Poor wording' }, { value: 'not_appropriate', label: 'Not academically appropriate' }, { value: 'other', label: 'Other' },
];

export interface SimilarQuestionMatch {
  existing_id: number | null;
  source: string;
  label: string | null;
  text: string;
  similarity_score: number;
  status: 'POTENTIAL_DUPLICATE' | 'HIGHLY_SIMILAR' | 'SOMEWHAT_SIMILAR' | 'NOT_SIMILAR';
}

export interface ConstraintChecks {
  topic: boolean | null;
  question_type: boolean;
  difficulty: boolean | null;
  cognitive_level: boolean | null;
  co_alignment: boolean | null;
  similarity: boolean;
  marks: boolean;
}

export interface QuestionValidation {
  detected_question_type: string | null;
  detected_difficulty: string | null;
  detected_cognitive_level: string | null;
  detected_topics: string[];
  co_alignment_score: number | null;
  co_alignment_status: 'STRONG' | 'WEAK' | 'NOT_ALIGNED' | null;
  max_similarity_score: number | null;
  similarity_status: SimilarQuestionMatch['status'] | null;
  similar_questions: SimilarQuestionMatch[];
  constraints: ConstraintChecks;
  warnings: string[];
  overall_status: ValidationStatus;
}

export interface GeneratedQuestion {
  id: number;
  generation_request_id: number;
  sequence: number;
  question_text: string;
  original_question_text: string;
  is_edited: boolean;
  question_type: GenQuestionType;
  marks: number;
  difficulty_level: GenDifficulty | null;
  cognitive_level: GenCognitive | null;
  learning_outcome_id: number | null;
  program_outcome_id: number | null;
  topic: string | null;
  options: string[] | null;
  correct_option: string | null;
  expected_answer: string | null;
  explanation: string | null;
  source_chunk_ids: number[];
  validation: QuestionValidation | null;
  validation_status: ValidationStatus;
  review_status: ReviewStatus;
  review_note: string | null;
  version: number;
  edited_at: string | null;
  approved_at: string | null;
  regenerated_from_id: number | null;
  official_question_id: number | null;
  added_to_assessment_at: string | null;
  can_add_to_assessment: boolean;
  created_at: string | null;
}

export interface BlueprintSlot {
  difficulty_level?: GenDifficulty | null;
  cognitive_level?: GenCognitive | null;
  question_type?: GenQuestionType | null;
  marks?: number | null;
  count: number;
}

export interface SetSummary {
  total: number;
  difficulty: Record<string, number>;
  cognitive_level: Record<string, number>;
  question_type: Record<string, number>;
  validation: Record<string, number>;
  potential_duplicates: number;
  weak_alignment: number;
  total_marks: number;
  co_coverage: Record<string, number>;
}

export interface GenerationRequest {
  id: number;
  course_id: number;
  assessment_id: number | null;
  course: { id: number; course_code: string | null; course_name: string | null } | null;
  assessment: { id: number; title: string; total_marks?: number | null; remaining_marks?: number | null } | null;
  learning_outcome: { id: number; code: string; description: string } | null;
  program_outcome: { id: number; code: string; title: string } | null;
  topic: string | null;
  question_type: GenQuestionType;
  difficulty_level: GenDifficulty | null;
  cognitive_level: GenCognitive | null;
  marks: number | null;
  number_of_questions: number;
  language: string;
  document_scope: { scope_type: string; course_id: number; document_id?: number | null; assessment_id?: number | null } | null;
  blueprint: BlueprintSlot[] | null;
  include_expected_answer: boolean;
  include_explanation: boolean;
  generation_status: GenerationStatus;
  generation_method: string | null;
  models: { generation: string | null; generation_version: string | null; embedding: string | null; prompt_version: string | null };
  regeneration_count: number;
  max_regenerations: number;
  feedback: string[];
  warnings: string[];
  set_summary: SetSummary | null;
  blueprint_summary: { requested: Record<string, Record<string, number>>; generated: Record<string, Record<string, number>>; matches: boolean } | null;
  retrieved_chunks: number;
  existing_questions_count: number;
  error_message: string | null;
  generated_questions_count: number | null;
  approved_count: number | null;
  disclaimer: string;
  completed_at: string | null;
  created_at: string | null;
  questions?: GeneratedQuestion[];
}

export interface CreateGenerationInput {
  course_id: number | string;
  assessment_id?: number | string | null;
  topic?: string | null;
  learning_outcome_id?: number | string | null;
  program_outcome_id?: number | string | null;
  question_type: GenQuestionType;
  difficulty_level?: GenDifficulty | null;
  cognitive_level?: GenCognitive | null;
  marks: number;
  number_of_questions?: number;
  language?: string;
  include_expected_answer?: boolean;
  include_explanation?: boolean;
  document_scope?: { scope_type: 'COURSE' | 'DOCUMENT' | 'ASSESSMENT'; document_id?: number | string | null; assessment_id?: number | string | null } | null;
  blueprint?: BlueprintSlot[] | null;
}

export interface UpdateGeneratedQuestionInput {
  question_text?: string;
  question_type?: GenQuestionType;
  marks?: number;
  difficulty_level?: GenDifficulty | null;
  cognitive_level?: GenCognitive | null;
  learning_outcome_id?: number | null;
  topic?: string | null;
  options?: string[] | null;
  correct_option?: string | null;
  expected_answer?: string | null;
  explanation?: string | null;
}

export interface FeedbackInput {
  feedback?: FeedbackReason[];
  feedback_note?: string;
}

File: ai-service/app/schemas/question_generation.py
"""STEP 33: Constrained Question Generator schemas."""

from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator

QuestionType = Literal["MCQ", "SHORT_ANSWER", "DESCRIPTIVE", "PROBLEM_SOLVING", "TRUE_FALSE", "CONCEPTUAL", "ANALYTICAL"]
DifficultyLevel = Literal["EASY", "MEDIUM", "HARD"]
CognitiveLevel = Literal["REMEMBER", "UNDERSTAND", "APPLY", "ANALYZE", "EVALUATE", "CREATE"]

QUESTION_TYPES = ["MCQ", "SHORT_ANSWER", "DESCRIPTIVE", "PROBLEM_SOLVING", "TRUE_FALSE", "CONCEPTUAL", "ANALYTICAL"]
DIFFICULTY_LEVELS = ["EASY", "MEDIUM", "HARD"]
COGNITIVE_LEVELS = ["REMEMBER", "UNDERSTAND", "APPLY", "ANALYZE", "EVALUATE", "CREATE"]


class CourseContext(BaseModel):
    course_code: Optional[str] = Field(None, max_length=50)
    course_name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=4000)


class OutcomeContext(BaseModel):
    code: str = Field(..., max_length=50)
    description: str = Field(..., max_length=4000)
    cognitive_level: Optional[str] = Field(None, max_length=30)


class ProgramOutcomeContext(BaseModel):
    code: str = Field(..., max_length=50)
    title: str = Field(..., max_length=255)
    description: Optional[str] = Field(None, max_length=4000)


class DocumentContextChunk(BaseModel):
    chunk_id: Optional[int] = None
    document_id: Optional[int] = None
    document_name: str = Field("Document", max_length=255)
    content: str = Field(..., min_length=1, max_length=20000)
    page_number: Optional[int] = None
    section_title: Optional[str] = Field(None, max_length=255)
    similarity_score: Optional[float] = None


class ExistingQuestion(BaseModel):
    id: Optional[int] = None
    text: str = Field(..., min_length=1, max_length=10000)
    source: str = Field("assessment", max_length=40)  # assessment | previous | question_bank
    label: Optional[str] = Field(None, max_length=255)


class BlueprintSlot(BaseModel):
    difficulty_level: Optional[DifficultyLevel] = None
    cognitive_level: Optional[CognitiveLevel] = None
    question_type: Optional[QuestionType] = None
    marks: Optional[float] = Field(None, gt=0, le=1000)
    count: int = Field(1, ge=1, le=20)


class GenerateQuestionsRequest(BaseModel):
    course_context: CourseContext = Field(default_factory=CourseContext)
    learning_outcome: Optional[OutcomeContext] = None
    program_outcome: Optional[ProgramOutcomeContext] = None
    topic: Optional[str] = Field(None, max_length=255)
    question_type: QuestionType = "DESCRIPTIVE"
    difficulty_level: Optional[DifficultyLevel] = None
    cognitive_level: Optional[CognitiveLevel] = None
    marks: float = Field(10, gt=0, le=1000)
    number_of_questions: int = Field(1, ge=1, le=20)
    language: str = Field("English", max_length=40)
    include_expected_answer: bool = True
    include_explanation: bool = False
    document_context: List[DocumentContextChunk] = Field(default_factory=list, max_length=20)
    existing_question_context: List[ExistingQuestion] = Field(default_factory=list, max_length=500)
    blueprint: Optional[List[BlueprintSlot]] = Field(None, max_length=20)
    feedback: List[str] = Field(default_factory=list, max_length=10)
    similarity_thresholds: Optional[Dict[str, float]] = None  # duplicate/high/moderate
    alignment_thresholds: Optional[Dict[str, float]] = None  # strong/weak

    @field_validator("topic")
    @classmethod
    def _strip_topic(cls, v: Optional[str]) -> Optional[str]:
        v = (v or "").strip()
        return v or None

    @field_validator("feedback")
    @classmethod
    def _clean_feedback(cls, v: List[str]) -> List[str]:
        return [f.strip()[:500] for f in v if f and f.strip()]


class ConstraintChecks(BaseModel):
    topic: Optional[bool] = None
    question_type: bool = True
    difficulty: Optional[bool] = None
    cognitive_level: Optional[bool] = None
    co_alignment: Optional[bool] = None
    similarity: bool = True
    marks: bool = True


class SimilarityMatch(BaseModel):
    existing_id: Optional[int] = None
    source: str
    label: Optional[str] = None
    text: str
    similarity_score: float
    status: str  # POTENTIAL_DUPLICATE | HIGHLY_SIMILAR | SOMEWHAT_SIMILAR | NOT_SIMILAR


class QuestionValidation(BaseModel):
    detected_question_type: Optional[str] = None
    detected_difficulty: Optional[str] = None
    detected_cognitive_level: Optional[str] = None
    detected_topics: List[str] = Field(default_factory=list)
    co_alignment_score: Optional[float] = None
    co_alignment_status: Optional[str] = None  # STRONG | WEAK | NOT_ALIGNED
    max_similarity_score: Optional[float] = None
    similarity_status: Optional[str] = None
    similar_questions: List[SimilarityMatch] = Field(default_factory=list)
    constraints: ConstraintChecks = Field(default_factory=ConstraintChecks)
    warnings: List[str] = Field(default_factory=list)
    overall_status: str = "PENDING"  # PASSED | PASSED_WITH_WARNINGS | FAILED


class GeneratedQuestionOut(BaseModel):
    question_text: str
    question_type: QuestionType
    marks: float
    difficulty_level: Optional[DifficultyLevel] = None
    cognitive_level: Optional[CognitiveLevel] = None
    topic: Optional[str] = None
    options: Optional[List[str]] = None
    correct_option: Optional[str] = None
    expected_answer: Optional[str] = None
    explanation: Optional[str] = None
    source_chunk_ids: List[int] = Field(default_factory=list)
    validation: QuestionValidation = Field(default_factory=QuestionValidation)


class BlueprintSummary(BaseModel):
    requested: Dict[str, Dict[str, int]] = Field(default_factory=dict)
    generated: Dict[str, Dict[str, int]] = Field(default_factory=dict)
    matches: bool = True


class GenerateQuestionsResponse(BaseModel):
    status: str = "success"
    questions: List[GeneratedQuestionOut]
    generation_method: str  # generative | template
    model: str
    model_version: str
    embedding_model: str
    prompt_version: str
    requested_count: int
    generated_count: int
    blueprint_summary: Optional[BlueprintSummary] = None
    warnings: List[str] = Field(default_factory=list)
    disclaimer: str
